arbnode/espresso_batcher_addr_monitor.go

```diff
@@ -0,0 +1,438 @@
+		events = append(events, event)
+	}
+	return events, nil
+}
+
+func (b *BatcherAddrMonitor) Store() error {
+	eventsBytes, err := rlp.EncodeToBytes(b.events)
+	if err != nil {
+		return fmt.Errorf("failed to encode events: %w", err)
+	}
+
+	initAddressesBytes, err := rlp.EncodeToBytes(b.initAddresses)
+	if err != nil {
+		return fmt.Errorf("failed to encode init addresses: %w", err)
+	}
+
+	newBatch := b.db.NewBatch()
+
+	err = newBatch.Put([]byte(eventKey), eventsBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put events: %w", err)
+	}
+
+	err = newBatch.Put([]byte(initAddressesKey), initAddressesBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put init addresses: %w", err)
+	}
+
+	return newBatch.Write()
+}
+
+func (b *BatcherAddrMonitor) Restore() error {
+	initAddressesBytes, err := b.db.Get([]byte(initAddressesKey))
+	if err != nil && !dbutil.IsErrNotFound(err) {
+		return fmt.Errorf("failed to get init addresses: %w", err)
+	}
+
+	if initAddressesBytes != nil {
+		err = rlp.DecodeBytes(initAddressesBytes, &b.initAddresses)
+		if err != nil {
+			return fmt.Errorf("failed to decode init addresses: %w", err)
+		}
+	}
+
+	eventsBytes, err := b.db.Get([]byte(eventKey))
+	if err != nil && !dbutil.IsErrNotFound(err) {
+		return fmt.Errorf("failed to get events: %w", err)
+	}
+
+	if eventsBytes != nil {
+		var events []BatcherAddrEvent
+		err = rlp.DecodeBytes(eventsBytes, &events)
+		if err != nil {
+			return fmt.Errorf("failed to decode events: %w", err)
+		}
+		b.events = events
+		b.cached = false
+		b.cachedAddresses = []common.Address{}
+		if len(events) > 0 {
+			b.lastEventL1Height = events[len(events)-1].L1Height
+		}
+	} else {
+		b.events = []BatcherAddrEvent{}
+		b.cached = false
+		b.cachedAddresses = []common.Address{}
+		b.lastEventL1Height = 0
+	}
+	return nil
+}
+
+func (b *BatcherAddrMonitor) backfill(ctx context.Context) error {
+	latestParentHeader, err := b.l1Reader.Client().HeaderByNumber(ctx, new(big.Int).SetInt64(int64(rpc.FinalizedBlockNumber)))
+	if err != nil {
+		return fmt.Errorf("failed to get latest parent height: %w", err)
+	}
+	currentParentHeight := b.GetConfirmedParentHeight()
+
+	if currentParentHeight == b.deployAt {
+		for _, addr := range b.initAddresses {
+			isBatcher, err := b.seqInboxInterface.IsBatchPoster(&bind.CallOpts{}, addr)
+			if err != nil {
+				return fmt.Errorf("failed to get batcher status: %w", err)
+			}
+			if !isBatcher {
+				return fmt.Errorf("init address %s is not a batcher", addr)
+			}
+		}
+
+		currentParentHeight = b.deployAt + 1
+		b.parentHeight = currentParentHeight
+	}
+
+	blocksToRead := uint64(100)
+	allowedRetry := 10
+	retry := 0
+	latestParentHeight := latestParentHeader.Number.Uint64()
+	log.Info("batcher addr monitor backfilling")
+	for retry < allowedRetry {
+		if currentParentHeight >= latestParentHeight {
+			break
+		}
+
+		events, err := b.LookupEvents(ctx, currentParentHeight+1, currentParentHeight+blocksToRead)
```

Comment:
the fromBlock should be the fromBlock we processed last time because its possible we miss some headers. So we should continuously store  the last processed block